src/arviz_plots/plots/utils.py

```diff
@@ -161,3 +166,70 @@ def set_grid_layout(pc_kwargs, plot_bknd, ds, num_rows=None, num_cols=None):
     pc_kwargs["plot_grid_kws"]["figsize"] = figsize
     pc_kwargs["plot_grid_kws"]["figsize_units"] = figsize_units
     return pc_kwargs
+
+
+def add_reference_lines(
+    plot_collection,
+    references,
+    orientation="vertical",
+    aes_map=None,
+    plot_kwargs=None,
+    backend=None,
+    data_vars=None,
+):
+    """Add reference lines.
+
+    This function adds lines to a plot collection based on the provided
+    references. It supports both vertical and horizontal lines, depending on the
+    specified orientation.
+
+    Parameters
+    ----------
+    plot_collection : PlotCollection
+        Plot collection to which the reference lines will be added.
+    references : list or dict
+        A list or dictionary of reference values to be plotted as lines.
+    orientation : str, default "vertical"
+        The orientation of the reference lines, either "vertical" or "horizontal".
+    aes_map : dict, optional
+        A dictionary mapping aesthetics to their corresponding variables.
+    plot_kwargs : dict, optional
+        A dictionary containing the plot arguments.
+    backend : str, optional
+        The backend used for plotting.
+    data_vars : list, optional
+        A list of data variables to be used for plotting.
+    """
+    if backend is None:
+        backend = plot_collection.backend
+    if plot_kwargs is None:
+        plot_kwargs = {}
+    if aes_map is None:
+        aes_map = {}
+    if data_vars is None:
+        data_vars = plot_collection.data.data_vars
+
+    plot_bknd = import_module(f".backend.{backend}", package="arviz_plots")
+
+    plot_func = vlines if orientation == "vertical" else hlines
+
+    _, ref_aes, ref_ignore = filter_aes(plot_collection, aes_map, "reference", "sample")
+    ref_kwargs = copy(plot_kwargs.get("reference", {}))
+
+    if "color" not in ref_aes:
+        ref_kwargs.setdefault("color", "gray")
+    if "linestyle" not in ref_aes:
+        ref_kwargs.setdefault("linestyle", plot_bknd.get_default_aes("linestyle", 2, {})[1])
+
+    if isinstance(references, dict):
+        for key, value in references.items():
+            if key in data_vars:
+                ref_dt = Dataset({key: DataArray(np.asarray(value))})
+                plot_collection.map(
+                    plot_func, "ref_lines", data=ref_dt, ignore_aes=ref_ignore, **ref_kwargs
+                )
+    else:
+        ref_dt = Dataset({var: DataArray(np.asarray(references)) for var in data_vars})
+        plot_collection.map(
+            plot_func, "ref_lines", data=ref_dt, ignore_aes=ref_ignore, **ref_kwargs
+        )
```

Comment:
Given both if branches generate datasets then call .map, I would move the call to .map outside the if/else block. In fact, I might even "outsource" this logic to a function in arviz-plots that given a dataset and some values in variable format like the references here it returns a new dataset filled with the provided values and similar/compatible dims+coords.

Rough idea:

```suggestion
    ref_dict = {}
    if isinstance(references, dict):
        for key, value in references.items():
            if key in data_vars:
                ref_dict[key] = DataArray(np.asarray(value))
    else:
        ref_dict = {var: DataArray(np.asarray(references)) for var in data_vars}
    plot_collection.map(
        plot_func, "ref_lines", data=Dataset(ref_dict), ignore_aes=ref_ignore, **ref_kwargs
    )
```

Main caveat not being taken into account. The fact that each plot can get multiple reference lines, which might end up as dataarrays with an extra `var_name_0` dimension instead of a shared one (even if it has nans in some positions).